panic when updating a CAA record

### Confirmation

- [X] My issue isn't already found on the issue tracker.
- [X] I have replicated my issue using the latest version of the provider and it is still present.

### Terraform and Cloudflare provider version

Terraform v1.0.7
on linux_amd64
+ provider registry.terraform.io/cloudflare/cloudflare v3.1.0


### Affected resource(s)

* `cloudflare_record`

### Terraform configuration files

```hcl
variable "config" {
  type = object({
    domain = string
    dns_records = any
  })
}

variable "zone_id" {
  type = string
}

locals {
  possible_a_records     = { for key, records in var.config.dns_records : key => records.a if records.a != null }
  possible_caa_records   = { for key, records in var.config.dns_records : key => records.caa if records.caa != null }
  possible_cname_records = { for key, records in var.config.dns_records : key => records.cname if records.cname != null }
  possible_mx_records    = { for key, records in var.config.dns_records : key => records.mx if records.mx != null }
  possible_ns_records    = { for key, records in var.config.dns_records : key => records.ns if records.ns != null }
  possible_txt_records   = { for key, records in var.config.dns_records : key => records.txt if records.txt != null }

  a_records = flatten([
    for key, records in local.possible_a_records : [
      for a_key, record in records : {
        state_identifier = "${key}_in_A_${a_key}"
        key              = key == "@" ? var.config.domain : key
        type             = "A"
        value            = record.address
        ttl              = record.ttl == "auto" ? 1 : record.ttl
        proxied          = record.proxied
        priority         = null
        caa_data         = null
      }
    ]
  ])

  caa_records = flatten([
    for key, records in local.possible_caa_records : [
      for caa_key, record in records : {
        state_identifier = "${key}_in_CAA_${caa_key}"
        key              = key == "@" ? var.config.domain : key
        type             = "CAA"
        value            = null
        ttl              = record.ttl == "auto" ? 1 : record.ttl
        proxied          = null
        priority         = null
        caa_data = {
          flags = record.flags
          tag   = record.tag
          value = record.value
        }
      }
    ]
  ])

  cname_records = [
    for key, record in local.possible_cname_records : {
      state_identifier = "${key}_in_CNAME"
      key              = key == "@" ? var.config.domain : key
      type             = "CNAME"
      value            = record.target
      ttl              = record.ttl == "auto" ? 1 : record.ttl
      proxied          = record.proxied
      priority         = null
      caa_data         = null
    }
  ]

  mx_records = flatten([
    for key, records in local.possible_mx_records : [
      for mx_key, record in records : {
        state_identifier = "${key}_in_MX_${mx_key}"
        key              = key == "@" ? var.config.domain : key
        type             = "MX"
        value            = record.mail_server
        ttl              = record.ttl == "auto" ? 1 : record.ttl
        proxied          = null
        priority         = record.priority
        caa_data         = null
      }
    ]
  ])

  ns_records = flatten([
    for key, records in local.possible_ns_records : [
      for ns_key, record in records : {
        state_identifier = "${key}_in_NS_${ns_key}"
        key              = key == "@" ? var.config.domain : key
        type             = "NS"
        value            = record.nameserver
        ttl              = record.ttl == "auto" ? 1 : record.ttl
        proxied          = null
        priority         = null
        caa_data         = null
      }
    ]
  ])

  txt_records = flatten([
    for key, records in local.possible_txt_records : [
      for txt_key, record in records : {
        state_identifier = "${key}_in_TXT_${txt_key}"
        key              = key == "@" ? var.config.domain : key
        type             = "TXT"
        value            = record.content
        ttl              = record.ttl == "auto" ? 1 : record.ttl
        proxied          = null
        priority         = null
        caa_data         = null
      }
    ]
  ])

  records = concat(
    local.a_records,
    local.caa_records,
    local.cname_records,
    local.mx_records,
    local.ns_records,
    local.txt_records
  )
}

resource "cloudflare_record" "records" {
  for_each = { for x in local.records : "${x.state_identifier}" => x }

  zone_id  = var.zone_id
  name     = each.value.key
  type     = each.value.type
  value    = each.value.value
  ttl      = each.value.ttl
  proxied  = each.value.proxied
  priority = each.value.priority

  // data for a CAA record
  dynamic "data" {
    for_each = each.value.caa_data != null ? [each.value.caa_data] : []
    content {
      flags = data.value["flags"]
      tag   = data.value["tag"]
      value = data.value["value"]
    }
  }
}
```


### Debug output

Too long for me to try redacting by hand, let me know if this is really necessary

### Panic output

Stack trace from the terraform-provider-cloudflare_v3.1.0 plugin:

panic: interface conversion: interface {} is []interface {}, not map[string]interface {}

goroutine 41 [running]:
github.com/cloudflare/terraform-provider-cloudflare/cloudflare.resourceCloudflareRecordUpdate(0xc000660900, {0xfdb3c0, 0xc0000fa0c0})
        github.com/cloudflare/terraform-provider-cloudflare/cloudflare/resource_cloudflare_record.go:510 +0xa39
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0xc0007198f0, {0x1206248, 0xc0009340c0}, 0x24, {0xfdb3c0, 0xc0000fa0c0})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.7.1/helper/schema/resource.go:352 +0x178
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc0000c0380, {0x1206248, 0xc0009340c0}, 0xc00015e3f0, 0xc0000abc20, {0xfdb3c0, 0xc0000fa0c0})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.7.1/helper/schema/resource.go:464 +0x6ba
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc00019c510, {0x1206248, 0xc0009340c0}, 0xc00071a3c0)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.7.1/helper/schema/grpc_provider.go:955 +0x9aa
github.com/hashicorp/terraform-plugin-go/tfprotov5/server.(*server).ApplyResourceChange(0xc0006d1860, {0x12062f0, 0xc00092e000}, 0xc0005887d0)
        github.com/hashicorp/terraform-plugin-go@v0.3.1/tfprotov5/server/server.go:332 +0x6c
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0xf9dbc0, 0xc0006d1860}, {0x12062f0, 0xc00092e000}, 0xc0007162a0, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.3.1/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:380 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000155a40, {0x1217c28, 0xc0005c6000}, 0xc000a98000, 0xc000265c80, 0x18b0e60, 0x0)
        google.golang.org/grpc@v1.40.0/server.go:1297 +0xccf
google.golang.org/grpc.(*Server).handleStream(0xc000155a40, {0x1217c28, 0xc0005c6000}, 0xc000a98000, 0x0)
        google.golang.org/grpc@v1.40.0/server.go:1626 +0xa2a
google.golang.org/grpc.(*Server).serveStreams.func1.2()
        google.golang.org/grpc@v1.40.0/server.go:941 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
        google.golang.org/grpc@v1.40.0/server.go:939 +0x294


### Expected output

Not to panic

### Actual output

Paniced, see panic output

### Steps to reproduce

1. import a CAA record with `terraform import`
2. Run `terraform apply`
3. Notice that the `terraform plan` attempts to re-add the `data` field:

```
  # module.example_com.cloudflare_record.records["example_in_CAA_primary"] will be updated in-place
  ~ resource "cloudflare_record" "records" {
      + allow_overwrite = false
        id              = "xxxxxxxxxxxxxxxxxxxxxx"
        name            = "example"
        # (10 unchanged attributes hidden)

      + data {
          + flags = "0"
          + tag   = "issue"
          + value = "certificate-authority.example.com"
        }

        # (1 unchanged block hidden)
    }

```
4. Accept the `apply`
5. Panic

### Additional factoids

_No response_

### References

_No response_

--- a/.changelog/1229.txt
+++ b/.changelog/1229.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+resource/cloudflare_record: handle `Update`s for records with `data` blocks
+```

--- a/cloudflare/resource_cloudflare_record.go
+++ b/cloudflare/resource_cloudflare_record.go
@@ -507,7 +507,8 @@ func resourceCloudflareRecordUpdate(d *schema.ResourceData, meta interface{}) er
 	newDataMap := make(map[string]interface{})
 
 	if dataOk {
-		for id, value := range data.(map[string]interface{}) {
+		dataMap := data.([]interface{})[0]
+		for id, value := range dataMap.(map[string]interface{}) {
 			newData, err := transformToCloudflareDNSData(updateRecord.Type, id, value)
 			if err != nil {
 				return err

--- a/cloudflare/resource_cloudflare_record_test.go
+++ b/cloudflare/resource_cloudflare_record_test.go
@@ -190,7 +190,16 @@ func TestAccCloudflareRecord_CAA(t *testing.T) {
 		CheckDestroy: testAccCheckCloudflareRecordDestroy,
 		Steps: []resource.TestStep{
 			{
-				Config: testAccCheckCloudflareRecordConfigCAA(rnd, zoneID, fmt.Sprintf("tf-acctest-caa.%s", domain)),
+				Config: testAccCheckCloudflareRecordConfigCAA(rnd, zoneID, fmt.Sprintf("tf-acctest-caa.%s", domain), 600),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckCloudflareRecordExists(resourceName, &record),
+					resource.TestCheckResourceAttr(resourceName, "data.0.flags", "0"),
+					resource.TestCheckResourceAttr(resourceName, "data.0.tag", "issue"),
+					resource.TestCheckResourceAttr(resourceName, "data.0.value", "letsencrypt.org"),
+				),
+			},
+			{
+				Config: testAccCheckCloudflareRecordConfigCAA(rnd, zoneID, fmt.Sprintf("tf-acctest-caa.%s", domain), 300),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckCloudflareRecordExists(resourceName, &record),
 					resource.TestCheckResourceAttr(resourceName, "data.0.flags", "0"),
@@ -636,7 +645,7 @@ resource "cloudflare_record" "%[3]s" {
 }`, zoneID, name, rnd)
 }
 
-func testAccCheckCloudflareRecordConfigCAA(resourceName, zoneID, name string) string {
+func testAccCheckCloudflareRecordConfigCAA(resourceName, zoneID, name string, ttl int) string {
 	return fmt.Sprintf(`
 resource "cloudflare_record" "%[1]s" {
   zone_id = "%[2]s"
@@ -647,8 +656,8 @@ resource "cloudflare_record" "%[1]s" {
     value = "letsencrypt.org"
   }
   type = "CAA"
-  ttl = 600
-}`, resourceName, zoneID, name)
+  ttl = %[4]d
+}`, resourceName, zoneID, name, ttl)
 }
 
 func testAccCheckCloudflareRecordConfigProxied(zoneID, domain, name, rnd string) string {

